CustomDialogOnClose/explainer.md

```diff
@@ -0,0 +1,125 @@
+   - The text that the website can display should be limited in length and not allow formatting.
+   - Always display "Leave" and "Cancel" buttons on the dialog, and only allow websites to add a third option which could be customizable.
+
+## Considered Alternatives
+### Extend window.confirm()
+[`window.confirm()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/confirm) is a familiar API that lets websites query the user and act based on their response.
+
+![A dialog is displayed with the title "An embedded page on this page says" and the contents "This is an example of the confirm API, this text is customizable". There are two buttons "OK" and "Cancel".](./images/ConfirmExample.PNG)
+
+We could extend this API to allow websites to add a new options to the dialog and have it return a promise with the user's selection. The website would then return a promise by calling .then() on window.confirm(). When the returned promise settles, the User Agent will close the tab. For example:
+    
+For example:
+```javascript
+window.addEventListener("beforeunload", (event) => {
+  if (featureDetectNewConfirmDialog()) {
+    return window.confirm("You have unsaved changes, would you like to save them?", "Save & Leave").then(function(result) {
+      // Perform the work to save the document.
+      resolve(); // The tab closes at this point.
+    });
+  } else {
+    event.preventDefault();
+  }
+});
+```
+
+However, this proposal brings confusion because window.confirm() would behave differently depending on if it was called from a beforeunload handler or not. It is also not a common pattern, and currently User Agents are allowed to deny showing window.confirm() dialogs from the beforeunload handler. For these reasons, this proposal was not chosen.
+
+### Fully customizable dialog
+
+We considered allowing the developer to design and show a fully customizable dialog, perhaps via the proposed [modal-window](https://github.com/adrianhopebailie/modal-window/blob/master/explainer.md), but this was not chosen due to it's additional complexity and the increased potential for abuse. For example, the site could choose to show only an "Accept" option, and no option to cancel. While mitigations for problems like this exist, the user experience would be worse than the proposed solution, where it is guaranteed that the user is given the option to leave a site immediately.
+
+## Privacy
+This capability does not have any meaningful privacy implications. It marginally increases the user's browser fingerprint by adding a feature detectable API, but this is minimal and is not a blocking issue.
+
+## Security
+This capability may have some security concerns by allowing malicious websites to trick users, as well as accepting untrusted input from web developers. Discussion is ongoing about how to mitigate the former, and a thorough security review of the code will be performed to ensure the latter is handled appropriately.
```

Comment:
There's some UI design mitigations that we should probably present somewhere as part of this discussion, either in text or in a screenshot. Some of these security issues arise from the fact that we're presenting text generated by a web developer in "official" browser UI. So this could have a similar treatment to `alert()` dialogs: clearly label text from the site (e.g. "this site says"). The tricky bit in that case would be how to handle the additional buttons. In all cases, design mitigations would be the purview of individual UAs, but good to mention here.

Reply:
I've added some text to the Avoiding Misuse section as well as the Security section to call out how UI design can help mitigate some of our security concerns.